src/mesh/coordinates.cxx

```diff
@@ -508,7 +510,40 @@ Coordinates::Coordinates(Mesh* mesh, Options* options, const CELL_LOC loc,
 
     /// Calculate Jacobian and Bxy
     if (jacobian()) {
-      throw BoutException("Error in jacobian call while constructing staggered Coordinates");
+      throw BoutException("Error in jacobian call while constructing staggered "
+                          "Coordinates");
+    }
+
+    // Attempt to read J from the grid file
+    Field2D Jcalc = J;
+    if (getAtLoc(mesh, J, "J", suffix, location)) {
+      output_warn.write(
+          "\tWARNING: Jacobian 'J_%s' not found. Calculating from metric tensor\n",
+          suffix);
+      J = Jcalc;
+    } else {
+      J = interpolateAndExtrapolate(J, location, extrapolate_x, extrapolate_y);
+
+      // Compare calculated and loaded values
+      output_warn.write("\tMaximum difference in J is %e\n", max(abs(J - Jcalc)));
+
+      // Re-evaluate Bxy using new J
+      Bxy = sqrt(g_22) / J;
+    }
+
+    // Attempt to read Bxy from the grid file
+    Field2D Bcalc = Bxy;
+    if (getAtLoc(mesh, Bxy, "Bxy", suffix, location)) {
+      output_warn.write("\tWARNING: Magnitude of B field 'Bxy_%s' not found. Calculating "
+                        " from metric tensor\n", suffix);
+      Bxy = Bcalc;
+    } else {
+      Bxy = interpolateAndExtrapolate(Bxy, location, extrapolate_x, extrapolate_y);
+
+      output_warn.write("\tMaximum difference in Bxy is %e\n", max(abs(Bxy - Bcalc)));
+      // Check Bxy
+      bout::checkFinite(Bxy, "Bxy" + suffix, "RGN_NOCORNERS");
+      bout::checkPositive(Bxy, "Bxy" + suffix, "RGN_NOCORNERS");
```

Comment:
shouldn't this be checked in all cases?
It would imho make sense to remove the check from `jacobian()` - as one might want to provide those values as the interpolation is causing issues.

Reply:
Yes, that makes a lot of sense.